### ISSUE TITLE
Class namespacing

### ISSUE THREAD

**Issue by hadley** [RConsortium/OOP-WG#4]
Should we explicitly include the class namespacing problem? i.e. with S3, both pkgA and pkgB can define a class `foo`, and there's no distinction at the dispatch level. In [adv-r](https://adv-r.hadley.nz/s3.html#s3-classes) I recommend using the package name as a class prefix to avoid this problem).

**Comment by lawremi** [comment]
Yes, we should. Part of having proper class definition objects is that we have all of that metadata.  Really there's no need to have a class assigned to a namespace. It could be that every class object is a reference object with a unique identity, independent of its name or anything else.

The question is how to attach the class object to instances. Simplest would be to make the class attribute an S3 object (extending character) with an attribute holding the class object.

**Comment by ltierney** [comment]
Class definition objects are a good idea. You have to keep in mind the issues of serialization/deserialization and also versioning and possibly patching up instances for version changes. CLOS has some work on this, as does Java. I don't know the details of either.

**Comment by wlandau** [comment]
How much would class namespacing help with #7?

**Comment by wlandau** [comment]
> Simplest would be to make the class attribute an S3 object (extending character) with an attribute holding the class object.

What if the attribute held a reference instead of the object itself? Or was that implied? Just wondering about performance. I have potential use cases in mind that would require upwards of 10k instances of the class at a time.

**Comment by lawremi** [comment]
Since it would be shallow copied, the object would already be attached by reference, in the form of a memory address. That should be faster than any explicit reference.

As a total aside, if you're working with 10k objects, you might think about a different data model, depending on what you are doing with them.
